File: python3/word_pattern.py
class Solution:
    def wordPattern(self, pattern: str, string: str) -> bool:
        strl = string.split(" ")

        if len(pattern) != len(strl):
            return False

        pat_map = {}
        str_map = {}
        for idx in range(len(pattern)):
            # print(pattern[idx], string_list[idx])
            if (pattern[idx] not in pat_map) and (strl[idx] not in str_map):
                pat_map[pattern[idx]] = strl[idx]
                str_map[strl[idx]] = pattern[idx]
                # print(pat_map, str_map)
            elif pattern[idx] not in pat_map:
                return False
            elif strl[idx] not in str_map:
                return False
            else:
                if pat_map[pattern[idx]] != strl[idx] or str_map[strl[idx]] != pattern[idx]:
                    return False

        return True

File: golang/word_pattern.go
package main

import "strings"

/*
Space	: O(n)
Time	: O(n)
*/

func wordPattern(pattern string, s string) bool {
	d1 := map[byte]string{}
	d2 := map[string]byte{}
	n := len(pattern)
	strs := strings.Split(s, " ")

	if len(strs) != n {
		return false
	}

	for i := 0; i < n; i++ {
		if _, ok := d1[pattern[i]]; ok {
			if d1[pattern[i]] != strs[i] {
				return false
			}
		}
		if _, ok := d2[strs[i]]; ok {
			if d2[strs[i]] != pattern[i] {
				return false
			}
		}
		d1[pattern[i]] = strs[i]
		d2[strs[i]] = pattern[i]
	}

	return true
}
